draft-aegis-mimi-arch-latest.md

```diff
@@ -1,156 +1,389 @@
++ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +     |
+                                                                      |
++ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +     |
+'                       Domain B                                '     |
+'                                                               '     |
+' +-------+     +----------------------+     +----------------+ '     |
+' | Carol | --- | Application Server 3 | --- | MIMI Gateway B |-'-----+ 
+' +-------+     +----------------------+     +----------------+ '
+' +-------+     +----------------------+       |                '
+' | Dave  | --- | Application Server 4 |-------+                '
+' +-------+     +----------------------+                        '
+'                                                               '
++ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+~~~
+
+# Gateway Overview
+
+The MIMI Gateway is designed to be a hybrid of AS and DS functionality as
+defined by `MLS Architecture`. It has an internal interface for
+applications within its domain to federate with each other, as well as an
+external interface that allows for cross-domain federation. A gateway acts as a
+permissioned database of the following information:
+
+* Unique identifiers of clients that can be queried by application-defined tags.
+* Queryable queues of MLS Key Packages indexed by identifier, protocol version,
+cipher suite and a custom label.
+* MLS proposals, commits, welcome messages and ratchet trees indexed by MLS
+`group_id` and `epoch_id`.
+
+## Identity Providers
+
+An identity provider helps ensure a consistent AS behavior between federating
+MLS applications. Internally an identity provider MUST implement the following
+set of functionality. 
+
+* Declare support for a specific set of MLS credential types.
+* Validate an identity based on its public key, credential, the current set of
+  group context extensions, and opaque application context. If validation is
+  done in the context of a commit, group context extensions MUST include any
+  changes due to a GroupContextExtensions proposal.
+* Uniquely identify a client based on its identity to ensure that the same
+  identity is not added to a MLS group multiple times.
+* Determine if one identity is a valid successor of another in order to verify
+  if an Update proposal, or a remove proposal within an external commit should
+  be allowed. Unless otherwise specified, a client is a valid successor of
+  another client if their client identifiers are equal.
+
+An example interface of an identity provider is as follows:
+
+~~~
+struct {
+    HPKEPublicKey public_key;
+    Credential credential;
+} Identity;
+
+struct {
+    opaque context<V>
+} ApplicationContext;
+
+interface {
+    fn supported_credentials() -> [CredentialType];
+
+    fn validate(Identity identity, Extension group_context_ext[],
```

Comment:
By forcing the context that comes from the MLS session to be the group_context_ext arent we forcing all MLS extensions that want to contribute to validation contexts to insert their data into the group_context? If so is that intended?